Update the OPAQUE-3DH Instantiation section (#172)
cfrg/draft-irtf-cfrg-opaque: draft-irtf-cfrg-opaque.md

15. Output (secret_creds.client_private_key, server_public_key, export_key)
~~~

## Protocol Compilation at Login {#protocol-compilation}

OPAQUE links an OPRF with an AKE protocol for login authentication. These can be executed concurrently, and the running
AKE can use the output values of the credential retrieval process, which includes the OPRF.

OPAQUE exposes the following API for authentication:

> - `ke1, blind, client_secret = OPAQUEStart(client_identity, password, client_info)`

Author: No, and I wanted to discuss that. I like your suggestions, let's do it this way.

- `ke2, client_info, expected_client_mac, session_key = OPAQUEResponse(server_identity, server_private_key, server_public_key, record, credential_identifier, oprf_seed)`
- `ke3, server_info, session_key = OPAQUEClientFinalize(password, blind, client_secret, client_identity, server_identity, ke1, ke2)`
- `valid = OPAQUEServerFinalize(expected_client_mac, ke3)`

{{opaque-client}} and {{opaque-server}} specify the inner working of these functions.

The `client_identity` and `server_identity` values must be consistent in the credential retrieval and AKE functions.
E.g., if the client falls back to using the public keys as identities, then they must be the same in the AKE stage.
If the server is set up to serve multiple clients, the client_identity can be used to differentiate clients and could be
send along with KE1.

The `ke1`, `ke2`, and `ke3` output values are the protocol messages to be exchanged among the peers, and their format
is proper to each AKE and defined by them.

> The return value of `AKEServerFinalize()` indicates the client's authenticity to the server and therefore the validity of

Reviewer: The function names in this paragraph seem outdated?

`client_info` and `session_key`produced by `AKEResponse()`. If it returns `false`, then the client MUST NOT be trusted and `client_info` and `session_key` MUST NOT be used.

> The following values are internal state values and displayed here for exhaustivity, to be carried from one stage to the other:

Reviewer: Since these are internal values, perhaps we can hide them under a "client_state" and "server_state" structure? Otherwise, APIs might be tempted to export them for applications to use.

3. Output (ke1, blind, client_secret)
~~~

~~~
OPAQUEClientFinalize(password, blind, client_secret, client_identity, server_identity, ke1, ke2)

Input:
- password, an opaque byte string containing the client's password
- blind, the OPRF blinding scalar
- client_secret, the client's Diffie-Hellman secret share for the session
- client_identity, the optional encoded client identity, which is set to its public key if not specified
- server_identity, the optional encoded server identity, which is set to its public key if not specified
- ke1, a KE1 message structure
- ke2, a KE2 message structure

Output:
- ke3, a KE3 message structure
- server_info, optional application-specific information sent by the server
- session_key, the session's shared secret 

Steps:
1. Create Credentials creds with (client_identity, server_identity)
2. client_private_key, server_public_key, export_key = RecoverCredentials(password, blind, ke2.CredentialResponse)
3. ke3, server_info, session_key = AKEClientFinalize(client_secret, client_identity, client_private_key,
                                               server_identity, server_public_key, ke1, ke2)
4. Output (ke3, server_info, session_key)
~~~

> The client's state between the two functions, except the parameters, are:

Reviewer: As above, I would recommend folding all this into a "state" or "context" object.

- client_identity (n.b. must match the one used in registration, recovery, and ake)
- password
- blind
- client_secret
- ke1

### OPAQUE Server {#opaque-server}

~~~
OPAQUEResponse(server_identity, server_private_key, server_public_key, record, credential_identifier, oprf_seed)

Input:

> - server_identity, the optional encoded server identity, which is set to its public key if not specified

Reviewer: ```suggestion
- server_identity, the optional encoded server identity, which is set to its public key if nil
```

- server_private_key, the server's private key
- server_public_key, the server's public key
- record, the client's RegistrationUpload structure
- credential_identifier, an identifier that uniquely represents the credential being registered
- oprf_seed, the server-side seed of Nh bytes used to generate an oprf_key

Output:
- KE2, a KE2 structure
- client_info, optional and unauthenticated application-specific information sent by the client
- expected_client_mac, the pre-computed mac of the full transcript expected in KE3
- server_secret, the server's Diffie-Hellman secret share for the session

Steps:
1. response = CreateCredentialResponse(ke1.request, server_public_key, record, credential_identifier, oprf_seed)
2. ke2, client_info, expected_client_mac, session_key = AKEResponse(server_identity, server_private_key, client_identity, record.client_public_key, server_info, KE1, response)
3. Output (ke2, client_info, expected_client_mac, session_key)
~~~

> It is of upmost importance that the server does not use `session_key` if `expected_client_mac` is not verified against
> KE3 in `OPAQUEServerFinalize()`.

Reviewer: ```suggestion
Servers MUST NOT use `session_key` unless `OPAQUEServerFinalize` returns `true`.
```

~~~
OPAQUEServerFinalize(expected_client_mac, ke3) = ServerFinalize(expected_client_mac, ke3)
~~~

> By pre-computing the necessary values in the response stage, the server's state holds in the following variables

Reviewer: Make this a state/context object, too?

We use the parameters Npk and Nsk to denote the size of the public and private keys used
in the AKE instantiation.

This specification defines OPAQUE with the 3DH AKE. Alternative AKEs are presented in {{alternate-akes}}.

### OPAQUE-3DH Instantiation {#opaque-3dh}

OPAQUE-3DH is implemented using a suitable prime order group. All operations in
the key derivation steps in {{derive-3dh}} are performed in this group and
represented here using multiplicative notation. The output of OPAQUE-3DH is a
session secret `session_key` and export key `export_key`.

OPAQUE-3DH allows the server to send additional encrypted and authenticated application specific information.
An opaque byte string `server_info` is encrypted using a one-time-pad derived from Ke2 and then used as input to XOR
with the `server_info` plaintext.

> The parameters Npk and Nsk are set to be equal to the size of an element and

Reviewer: Should we reference the dependencies section, where this is specified?

Author: Maybe this phrase is not even necessary, we could drop it

scalar, respectively, in the associated prime order group.

#### Flow {#opaque-3dh-flow}

~~~
 Client (client_info)             Server (server_identity, server_private_key, client_identity, client_public_key, serverInfo)
 --------------------------------------------------------------------

>  KE1 = Start(client_info)

Reviewer: The "AKE" prefix is missing here. I think it reads better without it, so maybe we drop "AKE" elsewhere?

Author: Yes, let's drop it

: Optional application-specific information to exchange during the protocol encrypted
under key Ke2, defined below.

server_mac
: An authentication tag computed over the handshake transcript computed using Km2,
defined below.

~~~
struct {
  uint8 client_mac[Nm];
} KE3;
~~~

client_mac
: An authentication tag computed over the handshake transcript computed using
Km2, defined below.

#### OPAQUE-3DH Functions {#opaque-3dh-functions}

##### OPAQUE-3DH Message Protocol functions

~~~
AKEStart(credential_request, client_info)

Parameters:
- Nn, the nonce length

Input:
- credential_request, a CredentialRequest structure
- client_info, the optional client_info to send to the server. Note that it will not be encrypted, and only authenticated 

Output:
- ke1, a KE1 structure
- client_secret, the client's ephemeral secret key of the Diffie-Hellman key exchange

Steps:
1. client_nonce = random(Nn)

> 2. client_secret = RandomPrivateKey()

Reviewer: Maybe we can add "GenerateKeyPair" to the AKE dependency? For example:

```
GenerateKeyPair(): Return a randomly generate private and public key pair. This
can be implemented by generating a random private key `sk` and then computing
`pk = RecoverPublicKey(sk)`.
```

> 3. client_keyshare = RecoverPublicKey(ephemeral_client_secret_key)

Reviewer: `ephemeral_client_secret_key` is undefined?

Author: it's `client_secret`, of course 🤦

- session_key, the shared session secret

Steps:
1. ikm = ClientIKM(client_secret, client_private_key, KE2.server_keyshare, server_public_key)
2. preamble = Preamble(client_identity, KE1, server_identity, KE2.inner_ke2)
3. Km2, Km3, handshake_encrypt_key, session_key = DeriveKeys(ikm, preamble)
4. expected_server_mac = MAC(Km2, Hash(concat(preamble, KE2.enc_server_info))
5. If !ct_equal(KE2.server_mac, expected_server_mac),
     raise MacError
6. client_mac = MAC(Km3, Hash(concat(preamble, KE2.enc_server_info, expected_server_mac))
7. pad = Expand(handshake_encrypt_key, "EncryptionPad", len(KE2.enc_server_info))
8. server_info = xor(pad, enc_server_info)
9. session_key = DeriveSecret(prk, "session secret", preamble)
10. Create KE3 ke3 with client_mac
11. Output (ke3, server_info, session_key)
~~~

~~~
AKEServerFinalize(expected_client_mac, ke3)

Input:
- expected_client_mac, the mac over the complete transcript pre-computed in `AKEResponse()`
- ke3, a KE3 structure

Output:
- bool, the boolean indicating whether the mac in KE3 is valid

Steps:
1. Output !ct_equal(KE3.client_mac, expected_client_mac)
~~~

###### Helper functions {#opaque-3dh-helper-functions)

> Random AKE Secret Key Share Generation Function:
> - RandomPrivateKey(): Generate a new scalar in the field of the AKE group using a byte string of length `Nsk` bytes
>   sourced from a cryptographically secure pseudorandom number generator (CSPRNG).

Reviewer: ```suggestion
```

This should just go in the AKE dependency section, I think. (See my comment above about GenerateKeyPair.)

Author: Will do after rebase

~~~

> ServerIKM(server_secret, server_private_key, client_keyshare, client_public_key)

Reviewer: Can we use the same function for both client and server here? In my implementation, that's what I do.

from TLS 1.3 {{?RFC8446}}.

~~~
Expand-Label(Secret, Label, Context, Length) =
    Expand(Secret, CustomLabel, Length)
~~~

Where CustomLabel is specified as:

~~~
struct {
   uint16 length = Length;
   opaque label<8..255> = "OPAQUE-" + Label;
   opaque context<0..255> = Context;
} CustomLabel;

Derive-Secret(Secret, Label, Transcript-Hash) =
    Expand-Label(Secret, Label, Transcript-Hash, Nx)
~~~

Note that the Label parameter is not a NULL-terminated string.

##### OPAQUE-3DH Key Schedule {#derive-3dh}

~~~
Preamble(client_identity, KE1, server_identity, inner_ke2)

Input:
- client_identity,
- KE1,
- server_identity,
- inner_ke2,

Output:
- preamble, 

Steps:
1. preamble = concat("3DH",
                     I2OSP(len(client_identity), 2), client_identity,
                     KE1,
                     I2OSP(len(server_identity), 2), server_identity,
                     KE2.inner_ke2)
2. Output preamble
~~~

~~~
DeriveKeys(ikm, preamble)

Parameters:
- Nx, the output size of the Extract function

Input:
- ikm, input key material
- preamble, 

Output:
- Km2,
- Km3,
- handshake_encrypt_key,
- session_key,

Steps:
1. prk = Extract("", ikm)

> 2. handshake_secret = Derive-Secret(prk, "HandshakeSecret", Hash(preamble))

Reviewer: What is `Hash` here?

3. session_key = Derive-Secret(prk, "SessionSecret", Hash(preamble))
4. Km2 = Derive-Secret(handshake_secret, "ServerMac", "")
5. Km3 = Derive-Secret(handshake_secret, "ClientMac", "")
6. handshake_encrypt_key = Derive-Secret(handshake_secret, "HandshakeEnc", "")
7. Output (Km2, Km3, handshake_encrypt_key, session_key)
~~~

~~~

> Expand-Label(secret, label, context, length)

Reviewer: I don't think we need to spell this function out. I'd suggest deleting it.
